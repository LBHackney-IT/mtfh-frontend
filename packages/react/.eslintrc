{
  "extends": "@lbh-mtfh/eslint-config",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*/rollup.config.js"]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ]
  }
}
