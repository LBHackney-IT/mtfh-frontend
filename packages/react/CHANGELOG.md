# @lbh-mtfh/react

## 1.1.0

### Minor Changes

- 37b7b3e: Initial release

### Patch Changes

- Updated dependencies [37b7b3e]
  - @lbh-mtfh/system@1.1.0
  - @lbh-mtfh/test-utils@1.1.0
