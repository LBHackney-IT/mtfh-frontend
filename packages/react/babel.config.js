const babel = require("../../babel.config")

module.exports = babel
