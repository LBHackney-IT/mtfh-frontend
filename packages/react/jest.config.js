const baseConfig = require("../../jest.config")

baseConfig.setupFilesAfterEnv.push("<rootDir>/jest.setup.ts")

module.exports = {
  ...baseConfig,
}
