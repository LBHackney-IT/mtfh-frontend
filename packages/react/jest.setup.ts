import * as LBHFrontend from "lbh-frontend"

window.LBHFrontend = LBHFrontend
