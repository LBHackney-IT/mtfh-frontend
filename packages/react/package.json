{
  "name": "@lbh-mtfh/react",
  "version": "1.0.0",
  "description": "React component library of the LBH Design System",
  "author": "https://hackney.gov.uk/",
  "homepage": "https://github.com/LBHackney-IT/mtfh-frontend",
  "license": "MIT",
  "sideEffects": false,
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/esm/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/cjs/index.js",
      "default": "./dist/esm/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LBHackney-IT/mtfh-frontend.git"
  },
  "scripts": {
    "test": "jest --env=jsdom --passWithNoTests",
    "build": "rollup -c"
  },
  "bugs": {
    "url": "https://github.com/LBHackney-IT/mtfh-frontend/issues"
  },
  "devDependencies": {
    "@lbh-mtfh/eslint-config": "^1.0.0",
    "@rollup/plugin-commonjs": "^19.0.0",
    "@rollup/plugin-node-resolve": "^13.0.0",
    "@rollup/plugin-typescript": "^8.2.1",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.0.0",
    "@testing-library/user-event": "^13.1.9",
    "axe-core": "^4.2.3",
    "axios": "^0.21.1",
    "jest": "^27.0.6",
    "jest-axe": "^5.0.1",
    "node-sass": "^6.0.1",
    "postcss": "^8.3.5",
    "rollup": "^2.52.8",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-postcss": "^4.0.0"
  },
  "dependencies": {
    "@lbh-mtfh/system": "^1.0.0",
    "@lbh-mtfh/test-utils": "^1.0.0",
    "@mtfh/common": "https://github.com/LBHackney-IT/mtfh-frontend-common.git",
    "@radix-ui/react-polymorphic": "^0.0.12",
    "@reach/accordion": "^0.15.3",
    "@reach/dialog": "^0.15.3",
    "@types/lbh-frontend": "^1.0.0",
    "classnames": "^2.3.1",
    "generic-names": "^3.0.0",
    "govuk-frontend": "^3.13.0",
    "lbh-frontend": "^3.5.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-merge-refs": "^1.1.0",
    "swr": "^0.5.6",
    "use-breakpoint": "^2.0.1"
  },
  "peerDependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
}
