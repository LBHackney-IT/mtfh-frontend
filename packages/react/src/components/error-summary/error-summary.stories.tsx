import React from "react";

import { ErrorSummary } from "./error-summary";

export default {
  title: "Error Summary",
  component: ErrorSummary,
};

export const Base = () => {
  return (
    <ErrorSummary id="error-summary" title="There is a problem">
      <a href="#error-1">The date your passport was issued must be in the past</a>
      <a href="#errror-2">Enter a postcode, like AA1 1AA</a>
    </ErrorSummary>
  );
};
