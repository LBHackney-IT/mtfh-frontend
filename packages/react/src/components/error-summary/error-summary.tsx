import React, {
  ComponentProps,
  ReactElement,
  forwardRef,
  isValidElement,
  useEffect,
  useRef,
} from "react";
import type * as Polymorphic from "@radix-ui/react-polymorphic";
import cn from "classnames";
import mergeRefs from "react-merge-refs";

import { widthOverrides } from "../../utils";
import { List } from "../list";
import styles from "./styles.module.scss";

export interface ErrorSummaryProps {
  id: string;
  title: string;
  description?: string;
  reFocus?: number;
  children?:
    | ReactElement<ComponentProps<"a">>
    | null
    | Array<ReactElement<ComponentProps<"a">> | null>;
  override?: number;
}

export type ErrorSummaryComponent = Polymorphic.ForwardRefComponent<
  "div",
  ErrorSummaryProps
>;

export const ErrorSummary: ErrorSummaryComponent = forwardRef(
  function ErrorSummary(
    {
      as: ErrorSummaryComp = "div",
      id,
      title,
      description,
      className,
      children,
      reFocus,
      override,
      ...props
    },
    ref
  ) {
    const localRef = useRef<HTMLElement>(null);

    useEffect(() => {
      /* istanbul ignore else */
      if (localRef.current && window?.GOVUKFrontend?.ErrorSummary) {
        // eslint-disable-next-line no-new
        new window.GOVUKFrontend.ErrorSummary(localRef.current);
        localRef.current.scrollIntoView(true);
      }
    }, []);

    useEffect(() => {
      /* istanbul ignore else */
      if (localRef.current) {
        localRef.current.scrollIntoView(true);
      }
    }, [reFocus]);

    return (
      <ErrorSummaryComp
        ref={mergeRefs([localRef, ref])}
        className={cn(
          styles.govukErrorSummary,
          styles.lbhErrorSummary,
          widthOverrides(override),
          className
        )}
        aria-labelledby={id}
        role="alert"
        {...props}
      >
        <h2 className={styles.govukErrorSummaryTitle} id={id}>
          {title}
        </h2>
        {description || children ? (
          <div className={styles.govukErrorSummaryBody}>
            {description ? <p>{description}</p> : null}
            {children ? (
              <List className={styles.govukErrorSummaryList}>
                {React.Children.map(
                  children,
                  (child) => child && isValidElement(child) && <li>{child}</li>
                )}
              </List>
            ) : null}
          </div>
        ) : null}
      </ErrorSummaryComp>
    );
  }
);
