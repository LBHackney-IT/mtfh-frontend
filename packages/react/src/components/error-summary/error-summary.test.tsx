import React from "react";
import { render, screen } from "@testing-library/react";
import { testA11y } from "@lbh-mtfh/test-utils";

import { ErrorSummary } from "./error-summary";

test("it renders an alert", async () => {
  const { container } = render(
    <ErrorSummary id="error" title="Something went wrong" />
  );
  expect(screen.getByRole("alert")).toBeInTheDocument();
  await testA11y(container);
});

test("it shows a description without children", () => {
  render(
    <ErrorSummary
      id="error"
      title="Something went wrong"
      description="User error"
    />
  );
  expect(screen.getByText("User error")).toBeInTheDocument();
});

test("it creates a list of the children", async () => {
  const { container } = render(
    <ErrorSummary id="error" title="Something went wrong">
      <a href="#test">Test</a>
      <a href="#next">Next</a>
    </ErrorSummary>
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});
