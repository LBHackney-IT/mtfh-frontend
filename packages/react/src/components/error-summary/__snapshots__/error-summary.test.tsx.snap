// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it creates a list of the children 1`] = `
<div>
  <div
    aria-labelledby="error"
    class="govukErrorSummary lbhErrorSummary"
    role="alert"
  >
    <h2
      class="govukErrorSummaryTitle"
      id="error"
    >
      Something went wrong
    </h2>
    <div
      class="govukErrorSummaryBody"
    >
      <ul
        class="lbhList govukErrorSummaryList"
      >
        <li>
          <a
            href="#test"
          >
            Test
          </a>
        </li>
        <li>
          <a
            href="#next"
          >
            Next
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
